sql/src/main/java/io/crate/planner/projection/TopNProjection.java

```diff
@@ -119,19 +121,24 @@ public String toString() {
     }
 
     @Nullable
-    public static TopNProjection createIfNeeded(Integer limit,
-                                                int offset,
-                                                int numOutputs,
-                                                List<DataType> inputTypes) {
+    public static Projection createIfNeeded(Integer limit,
+                                            int offset,
+                                            int numOutputs,
+                                            List<DataType> inputTypes) {
         if (limit == null) {
             limit = TopN.NO_LIMIT;
         }
-        if (limit == TopN.NO_LIMIT && offset == 0 && numOutputs >= inputTypes.size()) {
-            return null;
+        int numInputTypes = inputTypes.size();
+        List<DataType> strippedInputs = inputTypes;
+        if (numOutputs < numInputTypes) {
+            strippedInputs = inputTypes.subList(0, numOutputs);
         }
-        if (numOutputs < inputTypes.size()) {
-            inputTypes = inputTypes.subList(0, numOutputs);
+        if (limit == TopN.NO_LIMIT && offset == 0) {
+            if (numOutputs >= numInputTypes) {
+                return null;
+            }
+            return new EvalProjection(InputColumn.fromTypes(strippedInputs));
```

Comment:
if this class could return a different class instance, I thinks this method must be moved away, maybe ProjectionBuilder?